components/function-controller/internal/controllers/serverless/deployment.go

```diff
@@ -31,176 +26,185 @@ const (
 	MinimumReplicasUnavailable = "MinimumReplicasUnavailable"
 )
 
-func (r *FunctionReconciler) isOnDeploymentChange(instance *serverlessv1alpha1.Function, rtmConfig runtime.Config, deployments []appsv1.Deployment, dockerConfig DockerConfig) bool {
-	expectedDeployment := r.buildDeployment(instance, rtmConfig, dockerConfig)
-	resourceOk := len(deployments) == 1 && r.equalDeployments(deployments[0], expectedDeployment, isScalingEnabled(instance))
+func stateFnCheckDeployments(ctx context.Context, r *reconciler, s *systemState) stateFn {
+	labels := s.internalFunctionLabels()
 
-	return !resourceOk
-}
+	r.err = r.client.ListByLabel(ctx, s.instance.GetNamespace(), labels, &s.deployments)
+	if r.err != nil {
+		return nil
+	}
 
-func (r *FunctionReconciler) onDeploymentChange(ctx context.Context, log logr.Logger, instance *serverlessv1alpha1.Function, rtmConfig runtime.Config, deployments []appsv1.Deployment, dockerConfig DockerConfig) (ctrl.Result, error) {
-	newDeployment := r.buildDeployment(instance, rtmConfig, dockerConfig)
+	if r.err = ctx.Err(); r.err != nil {
```

Comment:
To stop computation in case of cancellation or if the deadline is exceeded.
BTW we've discussed it with @dbadura and the idea was to move the check up e.g.
```
...
func (m *reconciler) reconcile(ctx context.Context, f serverlessv1alpha1.Function) (ctrl.Result, error) {
	state := systemState{instance: f}

loop:
	for m.fn != nil {
		select {
		case <-ctx.Done():
			m.err = ctx.Err()
			break loop
		default:
			m.fn = m.fn(ctx, m, &state)

		}
	}

	return m.result, m.err
}
```